crates/log_probe/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+    #[error("Network error: {0}")]
+    NetworkError(String),
+}
+
+pub struct LogProbe {
+    tx: SyncSender<AggregatorMessage>,
+    shutdown: Arc<AtomicBool>,
+}
+
+impl Drop for LogProbe {
+    fn drop(&mut self) {
+        self.shutdown.store(true, Ordering::Relaxed);
+    }
+}
+
+enum AggregatorMessage {
+    AddRecord(Level),
+    Tick,
+}
+
+impl LogProbe {
+    pub fn new(
+        server_address: Option<String>,
+        report_frequency: Duration,
+    ) -> Result<Self, LogProbeError> {
+        let shutdown = Arc::<AtomicBool>::default();
+        let client = if let Some(address) = server_address {
+            Client::new(address)
+        } else {
+            Client::try_default()
+        }
+        .map_err(|e| LogProbeError::NetworkError(format!("{e}")))?;
+
+        let (tx, rx) = sync_channel::<AggregatorMessage>(100);
+        spawn_aggregator_thread(rx, client, shutdown.clone());
+        spawn_tick_thread(tx.clone(), shutdown.clone(), report_frequency);
+
+        Ok(Self { tx, shutdown })
+    }
+}
+
+fn spawn_tick_thread(
+    tx: SyncSender<AggregatorMessage>,
+    shutdown: Arc<AtomicBool>,
+    report_frequency: Duration,
+) {
+    std::thread::spawn(move || {
+        let start = Instant::now();
+        for deadline in (0..).map(|i| start + i * report_frequency) {
+            if let Err(e) = tx.send(AggregatorMessage::Tick) {
+                eprintln!("Failed to communicate with aggregator thread: {e}")
+            }
+
+            if shutdown.load(Ordering::Relaxed) {
+                break;
+            }
+
+            std::thread::sleep(deadline - Instant::now());
+        }
+    });
+}
+
+fn spawn_aggregator_thread(
+    rx: Receiver<AggregatorMessage>,
+    client: Client,
+    shutdown: Arc<AtomicBool>,
+) {
+    let mut log_stats = LogStats::default();
+    let mut report_start = Instant::now();
+
+    std::thread::spawn(move || {
+        for message in rx {
+            if shutdown.load(Ordering::Relaxed) {
+                break;
+            }
+
+            match message {
+                AggregatorMessage::AddRecord(record) => add_record(&mut log_stats, record),
+                AggregatorMessage::Tick => {
+                    log_stats.span = report_start.elapsed();
+                    if let Err(err) = client.send(&composer_api::Event::LogStats(log_stats)) {
+                        eprintln!("Could not send event {:?}", err)
+                    };
+                    log_stats = Default::default();
+                    report_start = Instant::now();
+                },
+            }
+        }
+    });
+}
+
+fn add_record(log_stats: &mut LogStats, record_level: Level) {
+    match record_level {
+        Level::Error => log_stats.error_records += 1,
+        Level::Warn => log_stats.warn_records += 1,
+        Level::Info => log_stats.info_records += 1,
+        Level::Debug => log_stats.debug_records += 1,
+        Level::Trace => log_stats.trace_records += 1,
+    }
+}
```

Comment:
Curiosity question: how one decides between a standalone

```rust
fn add_record(log_stats: &mut LogStats, ...)
```

and

```rust 
impl LogStats {
    fn add_record(&mut self, ...)
}
```
?

My intuition from java/kotlin/typescript would be the later just for the (subjectively) nicer ergonomics at the call site.

Reply:
Definitely the latter for me generally :)

However, in this case the problem is that `LogStats` is a *foreign type* (it's defined in a separate crate) so you can't have an impl block here. The options then are:
* define the `add_record` function alongisde `LogStats`, which isn't good as it leaks the `log` dependency into the `composer_api` crate where it doesn't need to be.
* Create an extension trait  (LogTraitExt?) with the `add_record` function and implement it for `LogStats`.
* Use a standalone function.

If `add_record` were to appear in the public API of `log_probe`, I'd likely use the second option, and maybe export the extension trait in some sort of prelude if I had a lot of these. But since `add_record` is an internal function that doesn't need nice ergonomics, the simplest thing is to go with the standalone, I think.

Reply:
Ahhh, I see. Thanks for explaining in length!

Agree that a standalone `fn` is best in this case.